File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    
    let store = {};
    let longest = 0;
    let len = 0;
    let start = 0;
    
     s.split("").forEach((item, index) => {
        if (store[item] || store[item] == 0) {
            if (start <= store[item]) {
                len = index - start;
                start = store[item] + 1;
                store[item] = index;
            }
            else {
                store[item] = index;
                len = (index - start) + 1;
            } 
        } else {
            store[item] = index;
            len = (index - start) + 1;
        }

        if (len > longest) {
            longest = len;
        }
    })
    
    return longest;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
// TC: O(N) // N = size of string s
// MC: O(1)
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> freq(256, 0);
        int longest = 0;
        
        for(int L = 0, R = 0; R < s.size(); R++) {
            freq[s[R]]++;
            bool hasDup = freq[s[R]] > 1;
            while(hasDup && L < s.size()) {
                freq[s[L]]--;
                if(freq[s[L]] == 1) hasDup = false;
                L++;
            };
            longest = max(longest, R - L + 1);
        }
        return longest;
    }
};